google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionClient.java

```diff
@@ -0,0 +1,287 @@
+  private final ScheduledExecutorService executor;
+  private final DatabaseId db;
+
+  @GuardedBy("this")
+  private volatile long sessionChannelCounter;
+
+  SessionClient(
+      SpannerImpl spanner,
+      DatabaseId db,
+      ExecutorFactory<ScheduledExecutorService> executorFactory) {
+    this.spanner = spanner;
+    this.db = db;
+    this.executorFactory = executorFactory;
+    this.executor = executorFactory.get();
+  }
+
+  @Override
+  public void close() {
+    executorFactory.release(executor);
+  }
+
+  /** Create a single session. */
+  SessionImpl createSession() {
+    // The sessionChannelCounter could overflow, but that will just flip it to Integer.MIN_VALUE,
+    // which is also a valid channel hint.
+    final Map<SpannerRpc.Option, ?> options;
+    synchronized (this) {
+      options = optionMap(SessionOption.channelHint(sessionChannelCounter++));
+    }
+    Span span = SpannerImpl.tracer.spanBuilder(SpannerImpl.CREATE_SESSION).startSpan();
+    try (Scope s = SpannerImpl.tracer.withSpan(span)) {
+      com.google.spanner.v1.Session session =
+          spanner
+              .getRpc()
+              .createSession(db.getName(), spanner.getOptions().getSessionLabels(), options);
+      span.end();
+      return new SessionImpl(spanner, session.getName(), options);
+    } catch (RuntimeException e) {
+      TraceUtil.endSpanWithFailure(span, e);
+      throw e;
+    }
+  }
+
+  /**
+   * Asynchronously creates a batch of sessions and returns these to the given {@link
+   * SessionConsumer}. This method may split the actual session creation over several gRPC calls in
+   * order to distribute the sessions evenly over all available channels and to parallelize the
+   * session creation. The given {@link SessionConsumer} is guaranteed to eventually get exactly the
+   * number of requested sessions unless an error occurs. In case of an error on one or more of the
+   * gRPC calls, the consumer will receive one or more {@link
+   * SessionConsumer#onSessionCreateFailure(Throwable, int)} calls with the error and the number of
+   * sessions that could not be created.
+   *
+   * @param sessionCount The number of sessions to create.
+   * @param consumer The {@link SessionConsumer} to use for callbacks when sessions are available.
+   */
+  void asyncBatchCreateSessions(final int sessionCount, SessionConsumer consumer) {
+    // We spread the session creation evenly over all available channels.
+    int sessionCountPerChannel = sessionCount / spanner.getOptions().getNumChannels();
+    int remainder = sessionCount % spanner.getOptions().getNumChannels();
+    int numBeingCreated = 0;
+    synchronized (this) {
+      for (int i = 0; i < spanner.getOptions().getNumChannels(); i++) {
+        // Create one more session for the first X calls to fill up the remainder of the division.
+        int createCountForChannel = sessionCountPerChannel + (i < remainder ? 1 : 0);
```

Comment:
i would simplify this

sessionCountPerChannel = sessionCount / spanner.getOptions().getNumChannels();
remainder = sessionCount % spanner.getOptions().getNumChannels();
// just add remainder to the requested count for last or first channel. 
for ( channel i : channels) {
 if ( i == 0 OR i == (numChannels - 1)) {
   sessionCountRequested =  sessionCountPerChannel + remainder
 }
}

Evenly distributing is great but wont buy us much optimization. Simple code will definitely improve readability and avoid bugs.

Reply:
Done.
This change also has the advantage that requests for numSessions < numChannels will be executed in one rpc call instead of numSessions rpc calls.

Reply:
Thanks !!

it has a be a huge config mistake if customers have  "numSessions < numChannels"
Though this change guards against using all those number of channels.